src/material_library.cpp

```diff
@@ -0,0 +1,221 @@
+              << std::endl;
+    exit(1);
+  }
+  if (!hdf5_path_exists(file, datapath)) {
+    std::cerr << "The datapath " << datapath << " in " << file << " empty."
+              << std::endl;
+    exit(1);
+  }
+
+  // load materials
+  from_hdf5(file);
+};
+
+// Destructor
+MaterialLibrary::~MaterialLibrary(){};
+
+// convert convert a filename into path+filename (for pyne)
+std::string MaterialLibrary::get_full_filepath(char* filename) {
+  std::string file(filename);
+  return MaterialLibrary::get_full_filepath(file);
+}
+
+// convert convert a filename into path+filename (for pyne)
+std::string MaterialLibrary::get_full_filepath(std::string filename) {
+  // remove all extra whitespace
+  filename.erase(std::remove(filename.begin(), filename.end(), ' '),
+                 filename.end());
+  // use stdlib call
+  const char* full_filepath = realpath(filename.c_str(), NULL);
+  return std::string(full_filepath);
+}
+
+// see if file exists
+bool MaterialLibrary::check_file_exists(std::string filename) {
+  // from http://stackoverflow.com/questions/12774207/
+  // fastest-way-to-check-if-a-file-exist-using-standard-c-c11-c
+  std::ifstream infile(filename.c_str());
+  return infile.good();
+}
+
+// Append Material to the library from hdf5 file
+void MaterialLibrary::from_hdf5(std::string filename, std::string datapath,
+                                int protocol) {
+  const char* data_path = datapath.c_str();
+
+  if (!hdf5_path_exists(filename, data_path)) return;
+
+  int num_materials = get_length_of_table(filename, datapath);
+
+  for (int i = 0; i < num_materials; i++) {
+    pyne::Material mat;  // from file
+    // read new mat
+    mat.from_hdf5(filename, datapath, i, protocol);
+
+    append_to_nuclist(mat);
+    material_library[mat.metadata["name"].asString()] = mat;
+    matlist.insert(mat.metadata["name"].asString());
+  }
+}
+
+void MaterialLibrary::add_material(pyne::Material mat) {
+  std::string mat_name = mat.metadata["name"].asString();
+  std::pair<pyne::mat_iter, bool> ins = material_library.insert(
+      std::pair<std::string, pyne::Material>(mat_name, mat));
+
+  if (ins.second) {
+    append_to_nuclist(mat);
+    matlist.insert(mat.metadata["name"].asString());
+  }
+}
+
+void MaterialLibrary::del_material(pyne::Material mat) {
+  std::string mat_name = mat.metadata["name"].asString();
+  del_material(mat_name);
+}
+
+void MaterialLibrary::del_material(std::string mat_name) {
+  material_library.erase(mat_name);
+  matlist.erase(mat_name);
+}
+
+pyne::Material MaterialLibrary::get_material(std::string mat_name) {
+  pyne::mat_iter it = material_library.find(mat_name);
+  if (it != material_library.end()) {
+    return it->second;
+  } else {
+    return pyne::Material();
+  }
+}
+
+void MaterialLibrary::write_hdf5(std::string filename, std::string datapath,
+                                 std::string nucpath, int chunksize) {
+  // Turn off annoying HDF5 errors
+  H5Eset_auto2(H5E_DEFAULT, NULL, NULL);
```

Comment:
What errors? Should we make an issue about it?

Reply:
I don't know, I pick that up from `pyne::material`